File: src/rheology/flow_viscosity/options/Options.jl
module Options

import EarthBox.OptionTools: OptionState
import EarthBox.OptionTools: make_option_names

const option_ids = Dict{Symbol, Int}(
    :XCoordinateDependent => -1,
    :Isoviscous => 0,
    :Composite => 1,
    :Composite2 => 2, # This is used to maintain backward compatibility with old models
    :TemperatureDependentCouetteBenchmark => 4,
    :TemperatureDependentConvectionBenchmark => 5
)

const option_names = make_option_names(option_ids)

function get_options()::Dict{Int, OptionState}
    composite_option = OptionState(
        option_name=string(option_names.Composite),
        description=(
            "Effective viscosity includes Peierls creep, dislocation creep and diffusion creep "
            *"flow laws depending on user input parameters."
            )
        )
    Dict{Int, OptionState}(
        option_ids[option_names.XCoordinateDependent] =>
            OptionState(
                option_name=string(option_names.XCoordinateDependent),
                description=(
                    "Viscosity is a function of lateral position of marker and pressure gradient."
                    ),
            ),
        option_ids[option_names.Isoviscous] =>
            OptionState(
                option_name=string(option_names.Isoviscous),
                description="Viscosity is constant",
            ),
        option_ids[option_names.Composite] => composite_option,
        # Maintain backward compatibility with old models
        option_ids[option_names.Composite2] => composite_option,
        option_ids[option_names.TemperatureDependentCouetteBenchmark] =>
            OptionState(
                option_name=string(option_names.TemperatureDependentCouetteBenchmark),
                description="Temperature-dependent viscosity for Couette flow benchmark.",
            ),
        option_ids[option_names.TemperatureDependentConvectionBenchmark] =>
            OptionState(
                option_name=string(option_names.TemperatureDependentConvectionBenchmark),
                description="Temperature-dependent viscosity for convection in a box benchmark.",
            )
    )
end

end # module 

File: src/rheology/flow_viscosity/options/options.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
""" Classes used to manage marker flow law types.
"""
from typing import Callable
from dataclasses import dataclass
from earthbox.model import ModelData
from earthbox.utilities.option_tools import OptionCollection
from earthbox.utilities.option_tools import Option
from ..utils.update import update_marker_flow_viscosity_blankenbach89
from ..utils.update import update_marker_flow_viscosity_couette_flow_benchmark
from ..utils.update import update_marker_flow_viscosity_isoviscous
from ..utils.update import update_marker_flow_viscosity_x_dependent
from ..utils.composite_viscosity import update_marker_composite_viscosity



UpdateViscosityFuncType = Callable[[ModelData], None]


class UpdateViscosityFuncs:
    """ Class used to define flow viscosity update functions.
    """
    x_coordinate_dependent: UpdateViscosityFuncType
    isoviscous: UpdateViscosityFuncType
    composite: UpdateViscosityFuncType
    temperature_dependent_couette_benchmark: UpdateViscosityFuncType
    temperature_dependent_convection_benchmark: UpdateViscosityFuncType

    def __init__(self):
        self.x_coordinate_dependent = (
            update_marker_flow_viscosity_x_dependent
            )
        self.isoviscous = (
            update_marker_flow_viscosity_isoviscous
            )
        self.composite = (
            update_marker_composite_viscosity
            )
        self.temperature_dependent_couette_benchmark = (
            update_marker_flow_viscosity_couette_flow_benchmark
            )
        self.temperature_dependent_convection_benchmark = (
            update_marker_flow_viscosity_blankenbach89
            )


class FlowLawOption(Option):
    """ Class used to define a flow law option.
    """
    update_viscosity_func: UpdateViscosityFuncType | None

    def __init__(
            self,
            option_name: str | None = None,
            description: str | None = None,
            update_viscosity_func: UpdateViscosityFuncType | None = None,
            flow_type_ids_used_in_function: tuple[int,...] | None = None
    ):
        super().__init__(option_name, description)
        self.update_viscosity_func = update_viscosity_func
        self.flow_type_ids_used_in_function = flow_type_ids_used_in_function

@dataclass
class FlowLawOptionNames:
    """ Class used to define a collection of option names.
    """
    x_coordinate_dependent: str = 'XCoordinateDependent'
    isoviscous: str = 'Isoviscous'
    composite: str = 'Composite'
    temperature_dependent_couette_benchmark: str = (
        'TemperatureDependentCouetteBenchmark')
    temperature_dependent_convection_benchmark: str = (
        'TemperatureDependentConvectionBenchmark')


class FlowLawOptionCollection(OptionCollection):
    """ Class used to define a collection of options.
    """
    options: dict[int, FlowLawOption] | None
    option_names: FlowLawOptionNames
    update_viscosity_funcs: UpdateViscosityFuncs

    def __init__(
        self,
        name: str,
        options: dict[int, FlowLawOption] | None = None
    ):
        super().__init__(name)
        self.options = options
        self.option_names = FlowLawOptionNames()
        self.update_viscosity_funcs = UpdateViscosityFuncs()
